File: tests/cpp/test_flac3d.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

// External includes
#include <gtest/gtest.h>

// System includes
#include <algorithm>
#include <cstdint>
#include <cstdio>
#include <fstream>
#include <map>
#include <sstream>
#include <string>
#include <vector>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/formats/flac3d.hpp"
#include "meshioplusplus/region.hpp"

namespace {

using meshioplusplus::Mesh;
using meshioplusplus::Region;
using meshioplusplus::RegionKind;

// A hand-written fixture rather than a read of tests/python/meshes: this suite
// has no test-data path, and a file written by our own writer is a weak oracle
// for our own reader -- a consistently wrong convention survives a round trip.
constexpr const char* kGroupsAscii = R"(* GRIDPOINTS
G 1 0.0 0.0 0.0
G 2 1.0 0.0 0.0
G 3 0.0 1.0 0.0
G 4 0.0 0.0 1.0
G 5 1.0 1.0 1.0
* ZONES
Z T4 1 1 2 3 4
Z T4 2 2 3 4 5
* ZONE GROUPS
ZGROUP "solid" SLOT "Default"
 1 2
ZGROUP "empty" SLOT "5"
* FACES
F T3 1 1 2 3
F T3 2 2 3 4
* FACE GROUPS
FGROUP "bottom" SLOT "Default"
 2
)";

std::string write_fixture(const std::string& rText, const std::string& rSuffix) {
    const std::string path = mt::temp_path(rSuffix);
    std::ofstream out(path, std::ios::binary);
    out << rText;
    out.close();
    return path;
}

// Every cell region as `name -> entries`; regions are stored sorted by
// (kind, name, dim, tag), but a map keeps the assertions readable either way.
std::map<std::string, std::vector<std::int64_t>> cell_regions(const Mesh& rMesh) {
    std::map<std::string, std::vector<std::int64_t>> out;
    for (std::size_t i = 0; i < rMesh.NumRegions(); ++i) {
        const Region& r = rMesh.Region(i);
        if (r.mKind != RegionKind::Cell)
            continue;
        out[r.mName] = std::vector<std::int64_t>(r.Entries(), r.Entries() + r.NumEntries());
    }
    return out;
}

// Append a little-endian scalar to a byte string.
template <class T>
void put(std::string& rOut, T v) {
    rOut.append(reinterpret_cast<const char*>(&v), sizeof(T));
}

// The same mesh as kGroupsAscii, in the binary layout.
std::string binary_fixture() {
    std::string b;
    put<std::uint32_t>(b, 1375135718u);
    put<std::uint32_t>(b, 3u);
    put<std::uint32_t>(b, 5u);  // points
    const double pts[5][3] = {{0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1}, {1, 1, 1}};
    for (std::uint32_t i = 0; i < 5; ++i) {
        put<std::uint32_t>(b, i + 1);
        for (int c = 0; c < 3; ++c)
            put<double>(b, pts[i][c]);
    }
    // zones (ids 1..2, their own namespace)
    put<std::uint32_t>(b, 2u);
    const std::uint32_t zones[2][4] = {{1, 2, 3, 4}, {2, 3, 4, 5}};
    for (std::uint32_t z = 0; z < 2; ++z) {
        put<std::uint32_t>(b, z + 1);
        put<std::uint32_t>(b, 4u);
        for (int j = 0; j < 4; ++j)
            put<std::uint32_t>(b, zones[z][j]);
    }
    // zone groups: "solid"/Default over both, "empty"/5 over none
    put<std::uint32_t>(b, 2u);
    put<std::uint16_t>(b, 5u);
    b += "solid";
    put<std::uint16_t>(b, 7u);
    b += "Default";
    put<std::uint32_t>(b, 2u);
    put<std::uint32_t>(b, 1u);
    put<std::uint32_t>(b, 2u);
    put<std::uint16_t>(b, 5u);
    b += "empty";
    put<std::uint16_t>(b, 1u);
    b += "5";
    put<std::uint32_t>(b, 0u);
    // faces (ids 1..2, restarting)
    put<std::uint32_t>(b, 2u);
    const std::uint32_t faces[2][3] = {{1, 2, 3}, {2, 3, 4}};
    for (std::uint32_t k = 0; k < 2; ++k) {
        put<std::uint32_t>(b, k + 1);
        put<std::uint32_t>(b, 3u);
        for (int j = 0; j < 3; ++j)
            put<std::uint32_t>(b, faces[k][j]);
    }
    // face groups: "bottom"/Default over face id 2
    put<std::uint32_t>(b, 1u);
    put<std::uint16_t>(b, 6u);
    b += "bottom";
    put<std::uint16_t>(b, 7u);
    b += "Default";
    put<std::uint32_t>(b, 1u);
    put<std::uint32_t>(b, 2u);
    return b;
}

// Blocks come back faces-first, so the two faces are global 0..1 and the two
// zones global 2..3 -- and `bottom` names face id 2, i.e. global 1.
const std::map<std::string, std::vector<std::int64_t>> kExpected = {
    {"face:bottom:Default", {1}},
    {"zone:empty:5", {}},
    {"zone:solid:Default", {2, 3}},
};

// Two triangles (faces) and two tetrahedra (zones) over five points.
Mesh base_mesh() {
    Mesh m;
    meshioplusplus::NDArray pts(meshioplusplus::DType::Float64, {5, 3});
    const double raw[15] = {0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1};
    std::copy(raw, raw + 15, pts.As<double>());
    m.AssignPoints(std::move(pts));

    meshioplusplus::NDArray tri(meshioplusplus::DType::Int64, {2, 3});
    const std::int64_t traw[6] = {0, 1, 2, 1, 2, 3};
    std::copy(traw, traw + 6, tri.As<std::int64_t>());
    m.AddCellBlock("triangle", std::move(tri));

    meshioplusplus::NDArray tet(meshioplusplus::DType::Int64, {2, 4});
    const std::int64_t qraw[8] = {0, 1, 2, 3, 1, 2, 3, 4};
    std::copy(qraw, qraw + 8, tet.As<std::int64_t>());
    m.AddCellBlock("tetra", std::move(tet));
    return m;
}

Mesh region_mesh() {
    Mesh m = base_mesh();

    meshioplusplus::NDArray e1(meshioplusplus::DType::Int64, {1});
    e1.As<std::int64_t>()[0] = 1;
    m.AddRegion(Region("face:bottom:Default", RegionKind::Cell, std::move(e1)));

    meshioplusplus::NDArray e2(meshioplusplus::DType::Int64, {2});
    e2.As<std::int64_t>()[0] = 2;
    e2.As<std::int64_t>()[1] = 3;
    m.AddRegion(Region("zone:solid:Default", RegionKind::Cell, std::move(e2)));
    return m;
}

std::string slurp(const std::string& rPath) {
    std::ifstream in(rPath, std::ios::binary);
    std::ostringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

}  // namespace

TEST(Flac3dGroups, AsciiFixture) {
    const std::string path = write_fixture(kGroupsAscii, ".f3grid");
    const Mesh m = meshioplusplus::read_flac3d(path);
    std::remove(path.c_str());
    EXPECT_EQ(cell_regions(m), kExpected);
}

TEST(Flac3dGroups, BinaryFixture) {
    const std::string path = write_fixture(binary_fixture(), ".f3grid");
    const Mesh m = meshioplusplus::read_flac3d(path);
    std::remove(path.c_str());
    // The two encodings of the same mesh must agree on the group names too:
    // the ascii slot arrives quoted and the binary one bare.
    EXPECT_EQ(cell_regions(m), kExpected);
}

TEST(Flac3dGroups, RoundTrip) {
    for (bool binary : {false, true}) {
        const Mesh in = region_mesh();
        const std::string path = mt::temp_path(".f3grid");
        meshioplusplus::write_flac3d(path, in, ".16e", binary);
        const Mesh out = meshioplusplus::read_flac3d(path);
        std::remove(path.c_str());
        EXPECT_EQ(cell_regions(out), cell_regions(in)) << "binary=" << binary;
    }
}

TEST(Flac3dGroups, NameIsAFixedPoint) {
    Mesh m = region_mesh();
    for (int pass = 0; pass < 2; ++pass) {
        const std::string path = mt::temp_path(".f3grid");
        meshioplusplus::write_flac3d(path, m, ".16e", false);
        m = meshioplusplus::read_flac3d(path);
        std::remove(path.c_str());
        EXPECT_EQ(cell_regions(m), cell_regions(region_mesh())) << "pass=" << pass;
    }
}

TEST(Flac3dGroups, FaceIdsRestartAtOne) {
    // Zones and faces are separate 1-based namespaces. Sharing one counter
    // made the ids disagree with the FGROUP lists written beside them.
    const std::string path = mt::temp_path(".f3grid");
    meshioplusplus::write_flac3d(path, region_mesh(), ".16e", false);
    const std::string text = slurp(path);
    std::remove(path.c_str());
    EXPECT_NE(text.find("\nZ T4 1 "), std::string::npos) << text;
    EXPECT_NE(text.find("\nF T3 1 "), std::string::npos) << text;
    EXPECT_NE(text.find("FGROUP \"bottom\" SLOT \"Default\""), std::string::npos) << text;
}

TEST(Flac3dGroups, NoRegionsWritesTheSameBytes) {
    // The guarantee that carrying groups costs a group-less mesh nothing: a
    // mesh with no Cell regions writes exactly the empty group sections it
    // always did.
    const Mesh without = base_mesh();

    const std::string a = mt::temp_path(".f3grid");
    meshioplusplus::write_flac3d(a, without, ".16e", false);
    const std::string text = slurp(a);
    std::remove(a.c_str());
    EXPECT_NE(text.find("* ZONE GROUPS\n* FACES"), std::string::npos) << text;
    EXPECT_NE(text.find("* FACE GROUPS\n"), std::string::npos) << text;
    EXPECT_EQ(text.find("ZGROUP"), std::string::npos) << text;
}

File: tests/python/test_flac3d.py
import pathlib

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus import _core
from meshioplusplus.flac3d import _flac3d

from . import helpers

THIS_DIR = pathlib.Path(__file__).resolve().parent
MESH_DIR = THIS_DIR / "meshes" / "flac3d"
REFERENCE_FILES = ["flac3d_mesh_ex.f3grid", "flac3d_mesh_ex_bin.f3grid"]

# The reference mesh's five groups, as *global* (block-major) cell indices.
# Recomputed from the raw file rather than from a meshio++ read: the ZGROUP /
# FGROUP member ids resolved through the file's own zone and face id tables.
# These are the numbers issue #76 reports, minus one for 0-basing.
REFERENCE_GROUPS = {
    "face:bottom:Default": (0, 18),
    "zone:Brick1:Default": (18, 45),
    "zone:Pyramid2:Default": (45, 72),
    "zone:Wedge3:Default": (72, 99),
    "zone:Tetrahedron4:Default": (99, 126),
}


def cell_regions(mesh):
    """``{name: entries}`` over the cell regions.

    A dict, not a list: the C++ core keeps regions sorted by
    ``(kind, name, dim, tag)`` while the Python ``Mesh`` keeps insertion
    order, so any cross-engine comparison has to be order-insensitive.
    """
    return {r.name: r.entries.tolist() for r in mesh.regions if r.kind == "cell"}


def expected_groups():
    return {k: list(range(*v)) for k, v in REFERENCE_GROUPS.items()}


# Both engines, explicitly: `meshioplusplus.read` prefers the C++ core, so a
# test that only goes through it says nothing about the numpy reference reader
# (verified -- sabotaging the Python rebase leaves such a test green).
ENGINES = {
    "core": lambda path: _core.flac3d_read(str(path)),
    "python": lambda path: _flac3d.read(str(path)),
}


@pytest.mark.parametrize(
    "mesh",
    [
        helpers.empty_mesh,
        helpers.tet_mesh,
        helpers.hex_mesh,
        helpers.tet_mesh,
        helpers.add_cell_sets(helpers.tet_mesh),
    ],
)
@pytest.mark.parametrize("binary", [False, True])
def test(mesh, binary, tmp_path):
    # mesh.write("out.f3grid")
    helpers.write_read(
        tmp_path,
        lambda f, m: meshioplusplus.flac3d.write(f, m, binary=binary),
        meshioplusplus.flac3d.read,
        mesh,
        1.0e-15,
    )


@pytest.mark.parametrize("filename", REFERENCE_FILES)
def test_reference_file(filename):
    filename = MESH_DIR / filename

    mesh = meshioplusplus.read(filename)

    # points
    assert np.isclose(mesh.points.sum(), 307.0)

    # cells
    ref_num_cells = [
        ("quad", 15),
        ("triangle", 3),
        ("hexahedron", 45),
        ("pyramid", 9),
        ("hexahedron", 18),
        ("wedge", 9),
        ("hexahedron", 6),
        ("wedge", 3),
        ("hexahedron", 6),
        ("wedge", 3),
        ("pyramid", 6),
        ("tetra", 3),
    ]
    assert [
        (cell_block.type, len(cell_block)) for cell_block in mesh.cells
    ] == ref_num_cells
    # Groups. Asserting the *membership*, not just the block count -- the old
    # `len(arr) == 12` check compared the number of blocks, which
    # `global_to_blocks` guarantees for any region, so it passed while three of
    # these five groups were empty and the other two were shifted (issue #76).
    assert cell_regions(mesh) == expected_groups()


@pytest.mark.parametrize("filename", REFERENCE_FILES)
@pytest.mark.parametrize("engine", sorted(ENGINES))
def test_reference_file_groups(filename, engine):
    """The membership the issue reports, asserted on each engine separately."""
    mesh = ENGINES[engine](MESH_DIR / filename)
    assert cell_regions(mesh) == expected_groups()


@pytest.mark.parametrize("filename", REFERENCE_FILES)
def test_reference_file_cell_sets_are_block_local(filename):
    """`cell_sets` holds per-block LOCAL indices; `regions` holds global ones.

    This is the convention the whole region layer is built on, and getting it
    wrong is what issue #76 was: the reader emitted global indices into the
    per-block slots, so `blocks_to_global` added the block base a second time
    and dropped everything that overshot its own block.
    """
    mesh = meshioplusplus.read(MESH_DIR / filename)

    blocks = mesh.cell_sets["face:bottom:Default"]
    assert [len(b) for b in blocks] == [15, 3] + [0] * 10
    assert blocks[0].tolist() == list(range(15))  # the 15 quads
    assert blocks[1].tolist() == list(range(3))  # the 3 triangles

    # Brick1 is the first 27 cells of the 45-hexahedron block, not the last.
    blocks = mesh.cell_sets["zone:Brick1:Default"]
    assert blocks[2].tolist() == list(range(27))


def test_ascii_and_binary_agree():
    """The same mesh in the two encodings must give the same group names.

    The ascii slot arrives quoted (`SLOT "Default"`) and the binary one bare,
    so without stripping, `face:bottom:"Default"` and `face:bottom:Default`
    were two different regions for the same group.
    """
    for read in ENGINES.values():
        assert cell_regions(read(MESH_DIR / REFERENCE_FILES[0])) == cell_regions(
            read(MESH_DIR / REFERENCE_FILES[1])
        )


@pytest.mark.parametrize("filename", REFERENCE_FILES)
@pytest.mark.parametrize("binary", [False, True])
def test_reference_file_round_trip(filename, binary, tmp_path):
    mesh = meshioplusplus.read(MESH_DIR / filename)

    path = tmp_path / "out.f3grid"
    meshioplusplus.flac3d.write(path, mesh, binary=binary)
    back = meshioplusplus.read(path)

    assert np.allclose(back.points, mesh.points)
    assert [(c.type, len(c)) for c in back.cells] == [
        (c.type, len(c)) for c in mesh.cells
    ]
    assert cell_regions(back) == expected_groups()


@pytest.mark.parametrize("binary", [False, True])
def test_group_names_are_a_fixed_point(binary, tmp_path):
    """`<zone|face>:<name>:<slot>` must survive repeated round trips.

    The writer used to hardcode the flag and slot, so each pass re-prefixed
    them: `zone:Brick1:Default` became `zone:zone:Brick1:Default:1`.
    """
    mesh = meshioplusplus.read(MESH_DIR / REFERENCE_FILES[0])
    names = sorted(cell_regions(mesh))

    for i in range(2):
        path = tmp_path / f"pass{i}.f3grid"
        meshioplusplus.flac3d.write(path, mesh, binary=binary)
        mesh = meshioplusplus.read(path)
        assert sorted(cell_regions(mesh)) == names


@pytest.mark.parametrize("filename", REFERENCE_FILES)
def test_cpp_matches_python_read(filename):
    path = str(MESH_DIR / filename)
    cpp = _core.flac3d_read(path)
    py = _flac3d.read(path)

    assert cell_regions(cpp) == cell_regions(py)
    assert np.allclose(cpp.points, py.points)
    assert [(c.type, len(c)) for c in cpp.cells] == [(c.type, len(c)) for c in py.cells]


@pytest.mark.parametrize("binary", [False, True])
def test_cpp_matches_python_write(binary, tmp_path):
    """Both engines must produce the same bytes, groups included."""
    mesh = _flac3d.read(str(MESH_DIR / REFERENCE_FILES[0]))

    cpp_path = tmp_path / "cpp.f3grid"
    py_path = tmp_path / "py.f3grid"
    _core.flac3d_write(str(cpp_path), mesh, ".16e", binary)
    _flac3d.write(py_path, mesh, binary=binary)

    assert cpp_path.read_bytes() == py_path.read_bytes()
    assert cell_regions(_core.flac3d_read(str(cpp_path))) == expected_groups()


@pytest.mark.parametrize("binary", [False, True])
def test_mixed_zone_and_face_mesh(binary, tmp_path):
    """Writing a mesh with both zones and faces used to raise outright.

    `split_f_z` zipped one category's block sizes against the whole cell list,
    and both sections shared one running cell-id counter even though a FLAC3D
    file numbers zones and faces independently.
    """
    mesh = meshioplusplus.Mesh(
        [[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]],
        [("tetra", [[0, 1, 2, 3]]), ("triangle", [[0, 1, 2]])],
        regions=[meshioplusplus.Region("mixed", "cell", [0, 1])],
    )

    path = tmp_path / "mixed.f3grid"
    meshioplusplus.flac3d.write(path, mesh, binary=binary)
    back = meshioplusplus.read(path)

    # The blocks come back faces-first, so the triangle is global 0.
    assert [c.type for c in back.cells] == ["triangle", "tetra"]
    # A region spanning both categories splits into an FGROUP and a ZGROUP.
    assert cell_regions(back) == {
        "face:mixed:Default": [0],
        "zone:mixed:Default": [1],
    }


def test_unknown_group_ids_are_dropped(tmp_path, capsys):
    """A group member the file never defines is dropped, never guessed at."""
    path = tmp_path / "unknown.f3grid"
    path.write_text(
        "* GRIDPOINTS\n"
        "G 1 0 0 0\n"
        "G 2 1 0 0\n"
        "G 3 0 1 0\n"
        "G 4 0 0 1\n"
        "* ZONES\n"
        "Z T4 1 1 2 3 4\n"
        "* ZONE GROUPS\n"
        'ZGROUP "g" SLOT "Default"\n'
        " 1 999\n"
    )
    mesh = _flac3d.read(path)
    assert cell_regions(mesh) == {"zone:g:Default": [0]}
    assert "does not define" in capsys.readouterr().err


@pytest.mark.parametrize("binary", [False, True])
def test_named_cell_set_round_trips(binary, tmp_path):
    """A bare set name is namespaced into the file's own vocabulary."""
    mesh = helpers.add_cell_sets(helpers.tet_mesh)
    n = len(mesh.cells[0])

    path = tmp_path / "sets.f3grid"
    meshioplusplus.flac3d.write(path, mesh, binary=binary)
    back = meshioplusplus.read(path)

    assert cell_regions(back) == {
        "zone:grain0:Default": list(range(0, n // 2)),
        "zone:grain1:Default": list(range(n // 2, n)),
    }
